core/src/main/scala/kafka/common/InterBrokerSender.scala

```diff
@@ -45,10 +45,15 @@ abstract class InterBrokerSendThread(
 
   private val unsentRequests = new UnsentRequests
 
-  def generateRequests(): Iterable[RequestAndCompletionHandler]
+  private val requestManagers = new ArrayList[InterBrokerRequestManager]()
```

Comment:
I think that would be handled in the managers right? I'm not sure what other synchronization we have here.

Reply:
Regarding synchronization, it seems that it's possible for `requestManagers` to be iterated through `drainGeneratedRequests()` while it's being modified through `addRequestManager()` concurrently. In that case, ArrayList could throw ConcurrentModificationException.

Regarding visibility, since there is no memory barrier for `requestManagers`, if one thread changes `requestManagers`, there is no guarantee that another thread will see the latest value inside `requestManagers`.

Reply:
Got it. Thanks for the explanation.

Reply:
Yeah. I think I just start the thread first and then add the managers. I'm ok with either approach.